src/timedisc/timedisc.jl

```diff
@@ -1,4 +1,155 @@
+  finalstep::Bool
+end
+
+function solve(ode::ODEProblem, alg::SimpleAlgorithm2N45;
+               dt, callback=nothing, kwargs...)
+  u = similar(ode.u0)
+  du = similar(u)
+  u_tmp = similar(u)
+  t = first(ode.tspan)
+  iter = 0
+  integrator = SimpleIntegrator2N(u, du, u_tmp, t, dt, zero(dt), iter, ode.p,
+                  (prob=ode,), alg,
+                  SimpleIntegrator2NOpts(callback, ode.tspan; kwargs...), false)
+  init!(integrator)
+  solve!(integrator)
+end
+
+function init!(integrator::SimpleIntegrator2N)
+  @unpack prob = integrator.sol
+  integrator.t = first(prob.tspan)
+  integrator.u .= prob.u0
+  integrator.iter = 0
+  integrator.finalstep = false
+
+  callbacks = integrator.opts.callback
+  if callbacks isa CallbackSet
+    for cb in callbacks.continuous_callbacks
+      error("unsupported")
+    end
+    for cb in callbacks.discrete_callbacks
+      cb.initialize(cb, integrator.u, integrator.t, integrator)
+    end
+  elseif !isnothing(callbacks)
+    error("unsupported")
+  end
+
+  return nothing
+end
+
+function solve!(integrator::SimpleIntegrator2N)
+  @unpack prob = integrator.sol
+  @unpack alg = integrator
+  t_end = last(prob.tspan)
+  callbacks = integrator.opts.callback
+
+  integrator.finalstep = false
+  @timeit_debug timer() "main loop" while !integrator.finalstep
+    if isnan(integrator.dt)
+      error("time step size `dt` is NaN")
+    end
+
+    # if the next iteration would push the simulation beyond the end time, set dt accordingly
+    if integrator.t + integrator.dt > t_end || isapprox(integrator.t + integrator.dt, t_end)
+      integrator.dt = t_end - integrator.t
+      terminate!(integrator)
+    end
+
+    # one time step
+    integrator.u_tmp .= 0
+    for stage in eachindex(alg.c)
+      t_stage = integrator.t + integrator.dt * alg.c[stage]
+      prob.f(integrator.du, integrator.u, prob.p, t_stage)
+
+      a_stage    = alg.a[stage]
+      b_stage_dt = alg.b[stage] * integrator.dt
+      @timeit_debug timer() "Runge-Kutta step" begin
+        Threads.@threads for i in eachindex(integrator.u)
+          integrator.u_tmp[i] = integrator.du[i] - integrator.u_tmp[i] * a_stage
+          integrator.u[i] += integrator.u_tmp[i] * b_stage_dt
+        end
+      end
+    end
+    integrator.iter += 1
+    integrator.t += integrator.dt
+
+    # handle callbacks
+    if callbacks isa CallbackSet
+      for cb in callbacks.discrete_callbacks
+        if cb.condition(integrator.u, integrator.t, integrator)
+          cb.affect!(integrator)
+        end
+      end
+    end
+  end
+
+  return (t=prob.tspan,
+          u=(copy(prob.u0), copy(integrator.u)))
+end
+
+get_du(integrator::SimpleIntegrator2N) = integrator.du
+
+u_modified!(integrator::SimpleIntegrator2N, ::Bool) = false
+
+function set_proposed_dt!(integrator::SimpleIntegrator2N, dt)
+  integrator.dt = dt
+end
+
+function terminate!(integrator::SimpleIntegrator2N)
+  integrator.finalstep = true
+  empty!(integrator.opts.tstops)
+end
```

Comment:
These methods should all get short comments

Reply:
Done in a0ca542.